TradeBeacons/data/scripts/systems/tradingoverview.lua

```diff
@@ -102,28 +102,64 @@ function getTradeBeaconScanRange(seed, rarity)
 end
 
 function requestSectorsData(caller)
-    if not caller then
+    local player
+    if caller then
+        -- triggered by player interaction so we can be specific
+        player = Player(caller)
+    elseif onServer() then
+        local faction = getParentFaction()
+        if faction.isPlayer then
+            -- only one choice for the player
+            player = Player(faction.index)
+        elseif faction.isAlliance then
+            -- TODO: maybe we can pick one in a more principled way?
+            --  or maybe trade beacons should be registering themselves with
+            --  trading systems instead of players?
+
+            local alliance = Alliance(faction.index)
+            local players = {Server():getOnlinePlayers()}
+            for _, p in pairs(players) do
+                if alliance:contains(p.index) then
+                    player = p
+                    break
+                end
+            end
+            if not player then
+                -- no player is logged in, so we don't have anywhere to get the
+                -- data from, no-op
+                return
+            end
+        else
+            -- faction isn't a player or alliance, no-op
+            return
+        end
+    else
+        -- no caller and not on server, this is unexpected, so no-op
         return
     end
 
     local entityId = Entity().index.string
     local x, y = Sector():getCoordinates()
     local script = "tradebeacon.lua"
 
-    Player(caller):invokeFunction(script, "requestSectorsData", x, y, historySize, entityId, caller)
+    player:invokeFunction(script, "requestSectorsData", x, y, historySize, entityId, caller)
 end
 
 function receiveTradingInfoFromPlayer(caller, sectorsDataString)
```

Comment:
this method now works a lot like collectSystemData used to. here is where we actually populate tradingData, and that happens independently from calling getData.

For the client, this still works because getData will trigger an (async) chain of calls that will end up here and the setData call below will trigger the trading UI to get refreshed with all the fancy data.

this does mean that getData will return stale data to the server. however, vanilla used to do the same thing (as stale as when last in the remote system), and the data can only be as fresh as the register calls from the beacons.

Reply:
I'm considering adding a requestSectorsData call to onSectorChanged which would make this even more similar to vanilla's update cadence. But the issue around choosing a Player to request the data from stopped me. See the above discussion.